File: prime.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <chrono>

using namespace std;
using namespace std::chrono;

ulong countChecks = 0;
const int TIME_LIMIT_SECONDS = 120;

bool is_prime(ulong n) {
    countChecks++;
    if (n <= 3)
        return n > 1;
    if (n % 2 == 0 || n % 3 == 0)
        return false;

    int i = 5;
    while (i * i <= n) {
        countChecks += 2;
        if (n % i == 0 || n % (i + 2) == 0)
            return false;
        i += 6;
    }
    return true;
}

void prime(ulong x, ulong y) {
    vector<ulong> primes;
    countChecks = 0;

    auto start = steady_clock::now();

    for (ulong num = max(2UL, x); num < y; ++num) {
        auto now = steady_clock::now();
        auto duration = duration_cast<seconds>(now - start).count();

        if (duration >= TIME_LIMIT_SECONDS) {
            cout << "\n⏱️ Time limit exceeded (" << TIME_LIMIT_SECONDS << "s). Stopping early.\n";
            break;
        }

        if (is_prime(num)) {
            primes.push_back(num);
        }
    }

    auto end = steady_clock::now();
    auto total_time = duration_cast<milliseconds>(end - start).count();

    cout << "Primes: ";
    for (ulong p : primes) {
        cout << p << " ";
    }
    cout << endl;

    cout << "Total primes number is: " << primes.size() << endl;
    cout << "Total checks: " << countChecks << endl;
    cout << "⏱️ Total time: " << total_time / 1000.0 << " seconds" << endl;
}

int main() {
    ulong  x, y;
    cout << "Type start number: ";
    cin >> x;
    cout << "Type last number: ";
    cin >> y;
    // y = static_cast<int>(pow(10, y));
    prime(x, y);
    return 0;
}

File: prime.py
#!/usr/bin/python3

import time
from sympy import isprime

TIME_LIMIT = 60  # giây
count = 0

def prime(x, y):
    global count
    count = 0
    primes = []

    start_time = time.time()

    for num in range(max(2, x), y):
        elapsed = time.time() - start_time
        if elapsed >= TIME_LIMIT:
            print(f"\n⏱️ Time limit exceeded ({TIME_LIMIT}s). Stopping early.")
            break

        if isprime(num):
            primes.append(num)
            count += 1  # Đếm số nguyên tố

    total_time = time.time() - start_time

    print("Primes:", primes)
    print("Total primes found:", len(primes))
    print("Total checks:", count)
    print(f"Total time: {total_time:.3f} seconds")

if __name__ == "__main__":
    x = int(input("Type start number: "))
    y = int(input("Type last number: "))
    
    prime(x, y)

